File: java/src/main/java/it/giannibombelli/wsc2026/payment/domain/events/RefundResultEvents.java
package it.giannibombelli.wsc2026.payment.domain.events;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;

public sealed interface RefundResultEvents extends PaymentEvent
    permits RefundResultEvents.TransactionRefunded, RefundResultEvents.TransactionNotRefunded {

    record TransactionRefunded(
        PaymentId aggregateId,
        ClientReference clientReference,
        Money amount
    ) implements RefundResultEvents {
        public TransactionRefunded {
            Require.requireArgument(aggregateId, "paymentId");
            Require.requireArgument(clientReference, "clientReference");
            Require.requireArgument(amount, "amount");
        }
    }

    record TransactionNotRefunded(
        PaymentId aggregateId,
        ClientReference clientReference,
        Description reason
    ) implements RefundResultEvents {
        public TransactionNotRefunded {
            Require.requireArgument(aggregateId, "paymentId");
            Require.requireArgument(clientReference, "clientReference");
            Require.requireArgument(reason, "reason");
        }
    }
}

File: typescript/src/payment/domain/events/refundResultEvents.ts
// Eventi di risultato del ciclo di vita di un rimborso.

import type { Event } from '@/common/domain/model/event.js';
import type { ClientReference } from '@/common/domain/primitive/clientReference.js';
import type { Description } from '@/common/domain/primitive/description.js';
import type { Money } from '@/common/domain/primitive/money.js';
import type { PaymentId } from '../payment/paymentId.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export type TransactionRefunded = Event<'TransactionRefunded', PaymentId> & {
  readonly clientReference: ClientReference;
  readonly amount: Money;
};

export function transactionRefunded(
  aggregateId: PaymentId,
  clientReference: ClientReference,
  amount: Money
): TransactionRefunded {
  requireArgument(aggregateId, 'paymentId');
  requireArgument(clientReference, 'clientReference');
  requireArgument(amount, 'amount');
  return { kind: 'TransactionRefunded', aggregateId, clientReference, amount };
}

export type TransactionNotRefunded = Event<'TransactionNotRefunded', PaymentId> & {
  readonly clientReference: ClientReference;
  readonly reason: Description;
};

export function transactionNotRefunded(
  aggregateId: PaymentId,
  clientReference: ClientReference,
  reason: Description
): TransactionNotRefunded {
  requireArgument(aggregateId, 'paymentId');
  requireArgument(clientReference, 'clientReference');
  requireArgument(reason, 'reason');
  return { kind: 'TransactionNotRefunded', aggregateId, clientReference, reason };
}

export type RefundResultEvent = TransactionRefunded | TransactionNotRefunded;
